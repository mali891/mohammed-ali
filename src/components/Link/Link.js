import React, { Fragment } from "react"
import PropTypes from "prop-types"
import classNames from "classnames"
import { Link as GatsbyLink } from "gatsby"

import { styles } from "./Link.styles"

const linkTypes = ["gatsby", "anchor", "button"]

const Link = ({ children, className, newTab, noHover, to, type, ...other }) => {
  const dynamicClassNames = classNames([className], {
    "c-link": true,
    "c-link--no-hover": noHover,
  })

  return (
    <Fragment>
      {type === "gatsby" && (
        <GatsbyLink
          css={styles}
          className={dynamicClassNames}
          activeClassName="c-link--active"
          to={to}
          {...other}
        >
          {children}
        </GatsbyLink>
      )}

      {type === "anchor" && (
        <a
          css={styles}
          className={dynamicClassNames}
          href={to}
          target={newTab ? "_blank" : null}
          rel={newTab ? "noopener noreferrer" : null}
          {...other}
        >
          {children}
        </a>
      )}

      {type === "button" && (
        <button css={styles} className={dynamicClassNames} {...other}>
          {children}
        </button>
      )}
    </Fragment>
  )
}

Link.defaultProps = {
  newTab: false,
  noHover: false,
  type: "gatsby",
}

Link.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]),
  className: PropTypes.string,
  newTab: PropTypes.bool,
  noHover: PropTypes.bool,
  to: PropTypes.string,
  type: PropTypes.oneOf(linkTypes),
}

export default Link
