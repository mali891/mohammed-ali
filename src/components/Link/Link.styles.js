import { css } from '@emotion/core'
import { COLOURS, FONT_FAMILY, EASINGS, BORDER_RADIUS } from '../../styles/settings'

export const styles = css`
	font-family: ${FONT_FAMILY.MEDIUM};
	color: ${COLOURS.LIGHT};
	background: none;
	border: none;
	text-decoration: none;
	position: relative;
	cursor: pointer;

	&:not(.c-link--no-hover) {
		&:hover {
			&:after {
				width: 100%;
			}
		}

		&:after {
			content: '';
			width: 0;
			position: absolute;
			left: 0;
			bottom: -1rem;
			right: 0;
			height: 0.4rem;
			width: 0;
			transition: all ${EASINGS.QUICK};
			background-color: ${COLOURS.LIGHT};
			border-radius: ${BORDER_RADIUS.LG};
		}
	}

	&.c-link--active {
		&:not(.c-link--no-hover) {
			&:after {
				width: 100%;
			}
		}
	}
`
