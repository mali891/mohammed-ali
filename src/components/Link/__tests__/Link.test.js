import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'

import Link from '../Link'

describe('Link', () => {
  it('renders correctly', () => {
    const { asFragment } = render(<Link />)
    expect(asFragment()).toMatchSnapshot()
  })

  it('renders correctly given a newTab prop', () => {
    const { asFragment, container } = render(<Link type="anchor" newTab />)

    expect(container.querySelector('a[target="_blank"]')).toBeInTheDocument()
    expect(asFragment()).toMatchSnapshot()
  })

  it('renders correctly given a noHover prop', () => {
    const { asFragment, container } = render(<Link noHover />)

    expect(container.querySelector('.c-link--no-hover')).toBeInTheDocument()
    expect(asFragment()).toMatchSnapshot()
  })

  it('renders correctly when given a specific to prop', () => {
    const { asFragment, container } = render(<Link type="anchor" to="test.com" />)

    expect(container.querySelector('a[href="test.com"]')).toBeInTheDocument()
    expect(asFragment()).toMatchSnapshot()
  })

  it('renders correctly when given a gatsby type prop', () => {
    const { asFragment } = render(<Link type="gatsby" />)

    expect(asFragment()).toMatchSnapshot()
  })

  it('renders correctly when given an anchor type prop', () => {
    const { asFragment, container } = render(<Link type="anchor" />)

    expect(container.querySelector('a')).toBeInTheDocument()
    expect(asFragment()).toMatchSnapshot()
  })

  it('renders correctly when given a button type prop', () => {
    const { asFragment, container } = render(<Link type="button" />)

    expect(container.querySelector('button')).toBeInTheDocument()
    expect(asFragment()).toMatchSnapshot()
  })

  it('renders correctly when given an onClick prop', () => {
    const mockOnClickHandler = jest.fn()
    const { container } = render(<Link onClick={mockOnClickHandler} />)
    const link = container.querySelector('.c-link')

    expect(mockOnClickHandler).toHaveBeenCalledTimes(0)
    fireEvent.click(link)
    expect(mockOnClickHandler).toHaveBeenCalledTimes(1)
  })
})
