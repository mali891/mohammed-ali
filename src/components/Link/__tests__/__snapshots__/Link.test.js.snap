// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link renders correctly 1`] = `
<DocumentFragment>
  <a
    class="c-link u-clr--secondary"
    css="[object Object]"
  />
</DocumentFragment>
`;

exports[`Link renders correctly given a newTab prop 1`] = `
<DocumentFragment>
  <a
    class="c-link u-clr--secondary"
    css="[object Object]"
    rel="noopener noreferrer"
    target="_blank"
  />
</DocumentFragment>
`;

exports[`Link renders correctly given a noHover prop 1`] = `
<DocumentFragment>
  <a
    class="c-link u-clr--secondary c-link--no-hover"
    css="[object Object]"
  />
</DocumentFragment>
`;

exports[`Link renders correctly when given a button type prop 1`] = `
<DocumentFragment>
  <button
    class="c-link u-clr--secondary"
    css="[object Object]"
  />
</DocumentFragment>
`;

exports[`Link renders correctly when given a gatsby type prop 1`] = `
<DocumentFragment>
  <a
    class="c-link u-clr--secondary"
    css="[object Object]"
  />
</DocumentFragment>
`;

exports[`Link renders correctly when given a specific to prop 1`] = `
<DocumentFragment>
  <a
    class="c-link u-clr--secondary"
    css="[object Object]"
    href="test.com"
  />
</DocumentFragment>
`;

exports[`Link renders correctly when given an anchor type prop 1`] = `
<DocumentFragment>
  <a
    class="c-link u-clr--secondary"
    css="[object Object]"
  />
</DocumentFragment>
`;
